conan/internal/api/detect_api.py

```diff
@@ -122,6 +122,47 @@ def _get_e2k_architecture():
     }.get(platform.processor())
 
 
+def detect_gnu_libc(ldd="/usr/bin/ldd"):
+    if platform.system() != "Linux":
+        ConanOutput(scope="detect_api").warning("detect_gnu_libc() only works on Linux")
+        return None
+    try:
+        first_line = check_output_runner(f"{ldd} --version").partition("\n")[0]
+        if any(glibc_indicator in first_line for glibc_indicator in ["GNU libc", "GLIBC"]):
+            return first_line.split()[-1].strip()
+        else:
+            ConanOutput(scope="detect_api").warning(
+                f"detect_gnu_libc() did not detect glibc in the first line of output from '{ldd} --version': '{first_line}'"
+            )
+            return None
+    except Exception as e:
+        ConanOutput(scope="detect_api").warning(
+            f"Couldn't get the glibc version from the '{ldd} --version' command {e}"
+        )
+        return None
+
+
+def detect_musl_libc(ldd="/usr/bin/ldd"):
```

Comment:
Functionally looks good, but it would be necessary to understand the usage patterns, which can be made explicit in a test.

Maybe it is better just a single function that returns ``libc, version = detect_api.detect_libc()``, so ``libc`` gets the value ``glibc`` or ``musl``?

Reply:
That's definitely possible and seems useful. I think it might imply the scope applies to more than just Linux, however, since macOS, BSD*, and Windows are all going to have their own libc implementation. I guess the functionality could always be expanded in the future, but it is quite nice to be able to pass `ldd` as an argument, too.

Reply:
Yes, exactly, I think it can be expanded in the future in different forms like ``detect_api.detect_osx_libc()``, via a new argument, or just by adding functionality if we are sure it will not break.

Reply:
Alright, I've added a `detect_libc` function to be used instead of ones for the individual C libraries which returns a tuple of the name, i.e. `gnu` or `musl`, and the version.

I broke out the actual parsing logic into separate functions and added unit tests for this logic using the output of the `ldd` command.

I added a functional test for the `detect_libc` function.